Remotion/Core/Development.Analyzers/RMCORE0001_SafeContextAnalyzer.cs

```diff
@@ -0,0 +1,285 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Core.Development.Analyzers
+{
+  [DiagnosticAnalyzer(LanguageNames.CSharp)]
+  public class RMCORE0001_SafeContextAnalyzer : DiagnosticAnalyzer
+  {
+    public static readonly DiagnosticDescriptor AlternativeDescriptor = new DiagnosticDescriptor(
+        "RMCORE0001",
+        "Use SafeContext instead of typical API",
+        "'{0}' should not be used with SafeContext, use 'Remotion.Context.SafeContext.{1}' instead.",
+        "Usage",
+        DiagnosticSeverity.Warning,
+        true);
+
+    public static readonly DiagnosticDescriptor WarningDescriptor = new DiagnosticDescriptor(
+        "RMCORE0001",
+        "Use typical API with SafeContext",
+        "'{0}' should not be used with SafeContext by itself, use '{1}' to create a safe context boundary.",
+        "Usage",
+        DiagnosticSeverity.Warning,
+        true);
+
+    public override void Initialize (AnalysisContext context)
+    {
+      context.EnableConcurrentExecution();
+      context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+      context.RegisterCompilationStartAction(
+          compilationContext =>
+          {
+            var safeContextTypeSymbol = compilationContext.Compilation.GetTypeByMetadataName("Remotion.Context.SafeContext");
+            if (safeContextTypeSymbol == null)
+              return;
+
+            var tasksTaskSymbol = compilationContext.Compilation.GetTypeByMetadataName("System.Threading.Tasks.Task");
+            var parallelSymbol = compilationContext.Compilation.GetTypeByMetadataName("System.Threading.Tasks.Parallel");
+            var threadSymbol = compilationContext.Compilation.GetTypeByMetadataName("System.Threading.Thread");
+            var threadingTimerSymbol = compilationContext.Compilation.GetTypeByMetadataName("System.Threading.Timer");
+            var executionSymbol = compilationContext.Compilation.GetTypeByMetadataName("System.Threading.ExecutionContext");
+            var timingTimerSymbol = compilationContext.Compilation.GetTypeByMetadataName("System.Timers.Timer");
+
+            var listOfMethodsAndActions = new List<(IMethodSymbol, Action<SyntaxNodeAnalysisContext, ISymbol>)>();
+
+            if (tasksTaskSymbol != null)
```

Comment:
I think we can cut down on the verbosity here still. How about we sit down together and take a look?